### ISSUE TITLE
Enhancement: add a user directory in which we can put our own devices files

### ISSUE THREAD

**Issue by jmcollin78** [make-all/tuya-local#2290]
Your integration is first class and save my life so many times with TuYa devices...

But I have some specific devices files, customized for my needs, and each time I update the integration I need to put my specific files into custom_components/tuya_local/devices

If possible, I would like that tuya_local search also in a specific user directory (in /config/tuyalocal_devices for example). This user specific directory is persistant after a restart or reinstallation or update of tuyalocal.

**Comment by make-all** [comment]
This has been requested before, but is not something that is desirable in my opinion, as it would encourage hoarding rather than sharing, which goes against the spirit of open source software.
localtuya exists for those who want to configure something differently for themselves.

**Comment by jmcollin78** [comment]
Hello, thank you for your reply. I understand the point "hoarding rather than sharing" but in my case, I have proposed the device.yaml file and it is still not included into the tuya-local. So each time, I upgrade, I have to copy manually the files from a backup. Looking for your backlog I understand you have many many devices to integrate and I understand it is difficult to deal with so much issues.

So it could help for all people that are waiting for the official integration update.

EDIT: my proposal https://github.com/make-all/tuya-local/issues/2108 . I think you prefer MR than issue. Are you taking into account some issues with the device file in attachement ?

**Comment by make-all** [comment]
In this case, supplying logs is essential, as without them I have no way to evaluate the contribution.
If you submit as a PR, I can have more confidence that you understand development practice and will continue to maintain it, but for contributions as part of an issue my assumption is that I will end up responsible for any flaws so I need the information to check it for it to be prioritised.

**Comment by bdf0506** [comment]
I agree with the OP here. I’ve had instances where I’ve submitted PRs and the maintainer has asked me to make modifications to fit the project’s theme. That’s fine and all, but once I make the changes to get the PR accepted I typically keep my original file and overwrite the file in the downloaded instance. The result? I update this repo like twice a year. With a user directory, I would do updates to pretty much each release. Feel like a user directory is relatively small ask all things considered.

**Comment by jmcollin78** [comment]
> I agree with the OP here. I’ve had instances where I’ve submitted PRs and the maintainer has asked me to make modifications to fit the project’s theme. That’s fine and all, but once I make the changes to get the PR accepted I typically keep my original file and overwrite the file in the downloaded instance. The result? I update this repo like twice a year. With a user directory, I would do updates to pretty much each release. Feel like a user directory is relatively small ask all things considered.

This is typically my use case.

**Comment by carmelo42** [comment]
Hi,
Having a specific user directory would be great indeed.
@make-all can you please accept ?

**Comment by jmcollin78** [comment]
Hello @make-all ,

As said in the first post, I love your integration.

But every time you update the integration I need to copy my specific files. Moreover I see the number of files and the number of FR and PR exploding. I wonder if it just possible to add as much files there is Tuya device.

Are you sure this model is sustainable in the futur ? How many Mb will weight your integration after some years ?

In my opinion, you should maintain the great solution, give some example with the integration and give a repository somewhere where you can download the files for its own use. 

Thanks for your attention,

**Comment by kongo09** [comment]
Isn't this a little bit like a HACS case, where most users appreciate a centrally curated list of integrations (here: devices) held to a high quality standard, but there can also be custom devices that are side loaded, but where there is no quality control.

One way to do this is to use a device repository for Tuya Local and an option for a personal repository for side loading. The personal repository would at least make such personal device variants public in an open source sense. 

Just my 2cts

**Comment by BerserkeR-Git** [comment]
If you really want to side load your custom yamls, you can do so with docker volume binds or something similar? That's what I've been doing before some of my devices were added.

**Comment by jmcollin78** [comment]
> If you really want to side load your custom yamls, you can do so with docker volume binds or something similar? That's what I've been doing before some of my devices were added.

I guess this is not possible with the haos mode. Only on docker mode you can change things maybe.

**Comment by stevene1919** [comment]
hey @make-all,

i put together a working implementation of this in a branch on my fork that scans `/config/custom_tuya_devices` alongside the internal `devices/` directory:
https://github.com/stevene1919/tuya-local/tree/feat/user-custom-devices-directory

it only touches `helpers/device_config.py` (+50 lines, -17 lines):
- checks if `/config/custom_tuya_devices` exists and includes it in `available_configs()` and `get_config()`
- user configs in the custom folder take precedence if a matching filename exists
- keeps the context manager cleanup intact

this makes it easy to develop and test new device configs locally without having in-progress files wiped on hacs updates.

happy to open a PR or let you pull from the branch if you'd like to include it.

**Comment by jmcollin78** [comment]
Thank you @stevene1919,

Thank you for your work. I encourage @make-all  to accept this PR and solve the most important issue with this great integration. Each time I need to update Tuya local, I need to copy/paste the directory with my personal config files. This is really annoying

**Comment by stevene1919** [comment]
i based my idea on what other integrations have done for custom configurations do for the support of being wiped when upgraded via HACs. 

I agree this is a great project and i would like to have my custom configurations not wiped while testing and take advantage of updates.
